### PR TITLE
Fix share completion with provider and spaces

### LINKED ISSUE #7569: path completion of unc paths with spaces fails to quote resulting path
Steps to reproduce
------------------

Start a path to a command, in this case,

`cd \\svr\`

Press CTRL-SPACE (<-EDITED) to bring up completion suggestions, which list all the shares on '\\\svr\\'.

Select a share that contains a space (notice the highlight fails to clear off entries as you move through the list). 

Expected behavior
-----------------

quotes (') should be added around the UNC path which contains characters that need escaped or quoted.

Actual behavior
---------------

Immediately you will notice there is no quote characters being inserted into the completion, and this prevents completing the next step in the path.

Side note, Windows PowerShell 5.1 seems to have the same problem with UNC paths, but at least the highlight works correctly.

Environment data
----------------

```powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      6.0.3
PSEdition                      Core
GitCommitId                    v6.0.3
OS                             Microsoft Windows 10.0.17134
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### LINKED ISSUE #10002: tab completion of a provider path notation UNC share name is not possible
related to #7569 and #9881 

# Steps to reproduce
Attempt to tab complete a share name on a UNC path via a partially resolved path that includes the provider notation.

```powershell
get-content Microsoft.PowerShell.Core\FileSystem::\\rk-vs-fs1\ # press tab to cycle through the shares on the server.
```

# Expected behavior

I expected the same completions that I get with
```powershell
get-content \\rk-vs-fs1\ # press tab to cycle through the shares on the server.
```

# Actual behavior

No completions occur

# Analysis

Upon researching leads on #9881, it was noticed that this issue would occur in this code:
https://github.com/PowerShell/PowerShell/blob/981c990d747ae97f56a6c7474e5d2ef04bf237c5/src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs#L4119-L4142
This particular code does not attempt to resolve the path name first to eliminate the potential 'filesystem' provider notation, before checking for a UNC '\\server\share' pattern.

The 'filesystem' provider notation does not affect any other part of the tab completion logic, only the share name completion.

Do note, this code shown above appears to be responsible for #7569, as it fails to escape `$` or spaces in the share names (or any other characters that might be legal in the share name but have a special meaning to PowerShell if not escaped or correctly quoted).

### PR DESCRIPTION
**Pull request by MartinGC94**
# PR Summary
Stops the regex pattern for fileshares from (incorrectly) finding Windows device paths.
Makes the list item/tooltip text for fileshare completions consistent with normal path completion behavior.
Fixes tab completion for share names when specifying the filesystem provider like this: `Microsoft.PowerShell.Core\FileSystem::\\localhost\<Tab>` previously you would get no results.

Also fixes tab completion for file shares which are currently not quoted as they should when they contain spaces.
I have a question about adding a test share for this scenario. Can I just do it? Do I need to do anything special with the share ACL?

## PR Context
Fixes https://github.com/PowerShell/PowerShell/issues/7569
Fixes https://github.com/PowerShell/PowerShell/issues/10002

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 1]
```diff
@@ -4380,27 +4380,38 @@ internal static IEnumerable<CompletionResult> CompleteFilename(CompletionContext
             var quote = HandleDoubleAndSingleQuote(ref wordToComplete);
             var results = new List<CompletionResult>();
 
-            // First, try to match \\server\share
-            // support both / and \ when entering UNC paths for typing convenience (#17111)
-            var shareMatch = Regex.Match(wordToComplete, @"^(?:\\\\|//)([^\\/]+)(?:\\|/)([^\\/]*)$");
+            // Matches file shares with and without the provider name and with either slash direction.
+            // Avoids matching Windows device paths like \\.\CDROM0 and \\?\Volume{b8f3fc1c-5cd6-4553-91e2-d6814c4cd375}\
+            var shareMatch = Regex.Match(
```
[the remark this answers is not in the document]
Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:4385]: PathIsUnc will return false if the filesystem provider is specified like this: `Microsoft.PowerShell.Core.Filesystem::\\localhost\c$`. So we can't use it as is.  
Also, It returns true when pointing to a file which doesn't match the current design of only matching up to the share portion.

[the remark this answers is not in the document]
Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:4385]: I don't know how often people use it but when they do we shouldn't be removing it from the completion result. You can specify the filesystem provider with or without the module name so if we want to avoid Regex then we need 2 "StartsWith" tests on the string along with the UNC check and the check to see if it's a file or a fileshare that the UNC path is pointing to.  
Maybe I'm underestimating the impact of regex but I don't feel like replacing 1 regex pattern with 4 other checks is worth doing.

[the remark this answers is not in the document]
Comment (SteveL-MSFT) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:4385]: Since this completion may be used often, we should compile and store a static version of the regex.

File: test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1 [hunk 2]
```diff
@@ -1269,6 +1269,11 @@ class InheritedClassTest : System.Attribute
             $res.CompletionMatches[0].CompletionText | Should -BeExactly $afterTab
         }
 
+        It "Tab completion UNC path with filesystem provider" -Skip:(!$IsWindows) {
```
Comment (SteveL-MSFT) [inline test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:1272]: Since your regex explicit avoids device paths, should have a test for that.

**Review by SteveL-MSFT — DISMISSED** [review]

File: test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1 (hunk 2 above)
Comment (MartinGC94) [reply test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:1272]: I couldn't test it because the wildcard globbing was broken (see: https://github.com/PowerShell/PowerShell/pull/19442 ) but now that it has been fixed, I'll add the test.

**Comment by JamesWTruher** [comment]
The @PowerShell/wg-powershell-engine discussed this and believe this change to be proper from a behavioral perspective.

**Comment by iSazonov** [comment]
> Stops the regex pattern for fileshares from (incorrectly) finding Windows device paths.

So follow will do not work?
```powershell
Get-ChildItem -Path \\.\C:\<tab>
```

File: test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1 (hunk 2 above)
Comment (iSazonov) [reply test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:1923]: 1. Before the PR follow works:
```
$res = TabExpansion2 -inputScript 'Filesystem::\\.\'
# or $res = TabExpansion2 -inputScript '\\.\'
$res.CompletionMatches[0] -eq '\\.\ADMIN$'
```
After the PR the result is empty.
We can add the test.

2. To test with space in share we can create a test share like `net share "tst qqq"="c:\1 1"`

Comment (MartinGC94) [reply test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:1934]: I don't remember and I can't figure out what I was talking about still being broken back then. Tab completion for device paths seems to work fine at the moment. Explicitly excluding device paths from the regex capture seems to have been more of a fix for correctness than a fix for broken functionality.  
With the old regex it would see `\\.\CDROM0<Tab>`  as a network share and try to run that section but it obviously wouldn't work. With my new pattern it correctly skips the share completion logic but because there's no logic to complete device names there won't be any results. If you add a trailing backslash: `\\.\CDROM0\` then both regex patterns correctly skips the share section and because you are now searching for items inside `\\.\CDROM0` it will also return actual results (if there are any).

So in conclusion there's no test to add for the device path regex exclusion because there's no functional change here.

As for your comment @iSazonov I'm not getting any results with: `TabExpansion2 -inputScript 'Filesystem::\\.\'` on my PC with a retail build without these changes. As for the test for a share with spaces, I will look into that later.

Comment (iSazonov) [reply test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:1934]: My first example doesn't work now but if worked in previously we can add the test to avoid regressions.

Comment (MartinGC94) [reply test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:1934]: Test added for share with space so I think all outstanding issues here have been resolved.

**Comment by MartinGC94** [comment]
I guess creating the share requires admin rights. Do you have a preferred way for me to skip this test when not elevated, or should I just add the classic:
```
using namespace System.Security.Principal
!([WindowsPrincipal][WindowsIdentity]::GetCurrent()).IsInRole([WindowsBuiltInRole]::Administrator)
```

to the skip parameter for the test?

**Comment by iSazonov** [comment]
@MartinGC94 There are CIs with elevated rights. Please use:
```
Describe "....." -Tag @('CI', 'RequireAdminOnWindows',) {
```

**Review by iSazonov — APPROVED** [review]
